первая часть проекта

Vladislav5108 opened the pull request:
ну так как все по новой, тут пока нет комментариев.

Vladislav5108 committed:
Описание ваших изменений

goandval commented on .DS_Store:
Удалить

goandval commented on cmd/inventorytracker/main.go:
```diff
@@ -0,0 +1,6 @@
+package main
+
+func main() {
+
```
Где запуск приложения?

goandval commented on internal/.DS_Store:
Удалить или добавить в .gitignore

goandval commented on internal/domain/entity/category.go:
```diff
@@ -0,0 +1,20 @@
+
+import "errors"
+
+var ErrNameCategory = errors.New("имя категории не должно быть пустым")
```
Всё на английском должно быть

goandval replied:
Актуально

goandval replied:
Актуально

goandval commented on internal/domain/entity/product.go:
```diff
@@ -0,0 +1,38 @@
+type Product struct {
+	ID         int
+	Name       string
+	Price      float64
```
Деньги хранятся либо в виде int в копейках, либо в виде decimal. Почему - почитай про машинное округление и почему деньги надо хранить в целых числах

goandval commented on internal/domain/repository/category.go:
```diff
@@ -0,0 +1,12 @@
+type CategoryRepository interface {
+	CreateCategory(category domain.Category) (int, error) //Создание новой категории
+	GetByIDCategory(id int) (domain.Category, error)      //получение категории по ID
+	GetALLCategory() ([]domain.Category, error)           //получение списка всех категорий
```
GetALLCategories

goandval commented on internal/domain/repository/category.go:
```diff
@@ -0,0 +1,12 @@
+	CreateCategory(category domain.Category) (int, error) //Создание новой категории
+	GetByIDCategory(id int) (domain.Category, error)      //получение категории по ID
+	GetALLCategory() ([]domain.Category, error)           //получение списка всех категорий
+	UpDateCategory(category domain.Category) error        //обновление категории
```
Update а не UpDate

goandval commented on internal/domain/repository/category.go:
```diff
@@ -0,0 +1,12 @@
+	GetByIDCategory(id int) (domain.Category, error)      //получение категории по ID
+	GetALLCategory() ([]domain.Category, error)           //получение списка всех категорий
+	UpDateCategory(category domain.Category) error        //обновление категории
+	DeleteCategory(id int) error                          //удаление кактегории
```
Комменты бессмысленны, так как названия методов так и переводятся

goandval commented on internal/domain/repository/product.go:
```diff
@@ -0,0 +1,20 @@
+	ProductRepository
+
+	Add(product domain.Product) (int, error) // добавляем новый товар
+	UpDate(product domain.Product) error     // обновляем данные товара
```
Update

goandval commented on internal/infrastructure/db/postgres/category_repo.go:
```diff
@@ -0,0 +1,134 @@
+	 $1,
+	 $2
+	 )
+	 RETURNING id`
```
Это не читаемо, надо форматировать по другому
INSERT INTO category(name,description)
	 VALUES($1, $2 )
	 RETURNING id

goandval commented on internal/infrastructure/db/postgres/category_repo.go:
```diff
@@ -0,0 +1,134 @@
+
+	err := c.db.QueryRow(query, category.Name, category.Description).Scan(&id)
+	if err != nil {
+		return 0, fmt.Errorf("ошибка создания категории: %w", err)
```
На английском

goandval commented on internal/infrastructure/db/postgres/category_repo.go:
```diff
@@ -0,0 +1,134 @@
+		&category.Description)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return domain.Category{}, domain.ErrNotId
```
Это хорошо

goandval commented on internal/infrastructure/db/postgres/category_repo.go:
```diff
@@ -0,0 +1,134 @@
+		if errors.Is(err, sql.ErrNoRows) {
+			return domain.Category{}, domain.ErrNotId
+		}
+		return domain.Category{}, fmt.Errorf("ошибка чтения данных: %w", err)
```
Англ

goandval commented on internal/infrastructure/db/postgres/category_repo.go:
```diff
@@ -0,0 +1,134 @@
+	return categories, nil
+}
+
+func (c *CategoryRepo) UpDateCategory(category domain.Category) error {
```
Update а не UpDate

goandval commented on internal/infrastructure/db/postgres/product_repo.go:
```diff
@@ -0,0 +1,254 @@
+	case errors.Is(err, sql.ErrNoRows):
+		return domain.Product{}, domain.ErrProductNotFound
+	case err != nil:
+		return domain.Product{}, err
```
Ошибки оборачиваем

goandval commented on internal/infrastructure/db/postgres/product_repo.go:
```diff
@@ -0,0 +1,254 @@
+		return domain.Product{}, domain.ErrProductNotFound
+	case err != nil:
+		return domain.Product{}, err
+	}
```
Здесь свитч лишний, выше посмотри такая же ситуация правильно обрабатывается через вложенный if

goandval commented on internal/infrastructure/db/postgres/product_repo.go:
```diff
@@ -0,0 +1,254 @@
+	$4,
+	false
+	)
+	RETURNING id`
```
Отформатировать, хотя бы табы проставить

goandval reviewed (changes requested):
Все ошибки и оборачивания должны быть на английском

goandval commented on internal/domain/entity/category.go:
```diff
@@ -3,7 +3,8 @@ package domain
 
 var ErrNameCategory = errors.New("имя категории не должно быть пустым")
-var ErrNotId = errors.New("ID не существует")
+var ErrNotFoundCategory = errors.New("ID не существует")
```
На английском

goandval replied:
Актуально

goandval commented on internal/domain/entity/product.go:
```diff
@@ -6,6 +6,7 @@ import (
+	ErrDuplicateName    = errors.New("товар с таким именем уже существует")
 	ErrInvalidIdProduct = errors.New("id должен быть положительным")
 	ErrProductNotFound  = errors.New("товара нет ")
 	ErrName             = errors.New("имя не должно быть пустым")
```
Английский

goandval replied:
Актуально

goandval commented on internal/infrastructure/db/postgres/product_repo.go:
```diff
@@ -33,13 +33,12 @@ func (r *ProductRepo) GetByID(id int) (domain.Product, error) {
+		}
 	}
-	return product, nil
+	return product, err
```
Это ошибка. Ты либо возвращаешь объект и nil, либо nil/пустую структуру и ошибку. То есть одно из двух. Сразу два не возвращают. Почему? Потому что на вызывающей стороне будет проверка if err != nil, и на второе значение после этого никто не будет смотреть. Но ты при этом его возвращаешь, значит, ошибка где-то в твоей логике разработчика

goandval commented on internal/infrastructure/db/postgres/product_repo.go:
```diff
@@ -118,19 +117,8 @@ func (r *ProductRepo) GetByCategory(CategoryID int) ([]domain.Product, error) {
-	$4,
-	false
-	)
+	name,price,quantity,category_id,is_archived)
```
Тут пробелы добавь для читаемости, вот как я тебе текст пишу
Тебе,же,не,удобно,будет,читать,вот,так
А, удобней, вот, так, вот

goandval replied:
Актуально

Vladislav5108 committed:
feat: добавить интерфейсы use case

- Создан и написан main
- Создан конфиг
- были отредактированы некоторые файлы т.к. искал проблему почему база данных возвращала пустые значения. Добавил новое поле IsArchived по этой причине. а оказалось что в таблице не хватало колонки CreatedAt (
-

Vladislav5108 committed:
перевел на ангийский кастомные ошибки

goandval commented on cmd/main.go:
```diff
@@ -0,0 +1,68 @@
+	log.Println("DB host:", cfg.Postgres.Host)
+
+	db, err := postgres.NewDB(postgres.Config{
+		Host:     cfg.Postgres.Host,
+		Port:     cfg.Postgres.Port,
+		User:     cfg.Postgres.User,
+		Password: cfg.Postgres.Password,
+		DBName:   cfg.Postgres.DBName,
+		SSLMode:  cfg.Postgres.SSLMode,
+		Timeout:  cfg.Postgres.Timeout,
```
Чем куча таких строчек вида а=а, лучше передавай сразу структуру cfg.Postgres в метод NewDB

goandval replied:
Актуально

Vladislav5108 committed:
Написал юнит тесты для основных слоев. Создал миграции и макфайл. Проверил работу всех запросов

goandval commented on cmd/server.log:
Это что? Лог надо в терминал stdout чтоб писал а не в файл

goandval commented on internal/transport/myhttp/product_admin_handler.go:
```diff
@@ -0,0 +1,167 @@
+	if err != nil {
+		log.Printf("Add product error: %v", err)
+		status := http.StatusInternalServerError
+		if strings.Contains(err.Error(), "foreign key") {
```
err.Error() можно вызывать только в тестах. В коде на это опираться нельзя. Ошибки постгреса можно проверять через определённые константы с ошибками. НАпрмиер, ошибка foreign key может иметь код 22054(к примеру), и дальше ты через пакет pgx а не sql сделаешь запросы к БД и сможешь ошибку БД сделать errors.As в ошибку постгреса и у неё посмотреть этот код. Ну это как в идеале, и для этого тебе надо будет слой БД переписать с библиотеки sql на pgx. Поэтому пока можешь так не делать время не тратить, но эту строку удали с err.Error() - это прям нельзя делать вообще

goandval commented on internal/transport/myhttp/product_admin_handler.go:
```diff
@@ -0,0 +1,167 @@
+		return
+	}
+	if id <= 0 {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "id must be a positive number "})
```
Чем везде писать "error": "" и тому подобное, заведи типы данных структуры ответов - позитивных, негативных

Vladislav5108 committed:
Создан dockerfile, исправлен конфиг под докер, и ошибки с прошлого комита
